src/riscv/linux/init.c

```diff
@@ -0,0 +1,112 @@
+#include <string.h>
+
+#include <cpuinfo/internal-api.h>
+#include <cpuinfo/log.h>
+#include <linux/api.h>
+#include <riscv/linux/api.h>
+
+#define ALLOCATE_OR_FAIL(dest, count, size, description) \
+	dest = calloc(count, size); \
+	if (dest == NULL) { \
+		cpuinfo_log_error("failed to allocate %zu bytes for %zu %s.", \
+				  count * size, \
+				  count, \
+				  description); \
+		goto cleanup; \
+	}
+
+
+/* ISA structure to hold supported extensions. */
+struct cpuinfo_riscv_isa cpuinfo_isa;
+
+void cpuinfo_riscv_linux_init(void) {
+	struct cpuinfo_riscv_linux_processor* riscv_linux_processors = NULL;
+	struct cpuinfo_processor* processors = NULL;
+
+	/**
+	 * The interesting set of processors are the number of 'present'
+	 * processors on the system. There may be more 'possible' processors, but
+	 * processor information cannot be gathered on non-present processors.
+	 *
+	 * Note: For SoCs, it is largely the case that all processors are known
+	 * at boot and no processors are hotplugged at runtime, so the
+	 * 'present' and 'possible' list is often the same.
+	 *
+	 * Note: This computes the maximum processor ID of the 'present'
+	 * processors. It is not a count of the number of processors on the
+	 * system.
+	 */
+	const size_t max_processor_id = 1 +
+		cpuinfo_linux_get_max_present_processor(
+				cpuinfo_linux_get_max_processors_count());
+	if (max_processor_id == 0) {
```

Comment:
This is not possible unless addition of `1` above overflows

Reply:
Correct - unfortunately the behavior of `cpuinfo_linux_get_max_present_processor` is to return UINT32_MAX when it fails. A similar check is used in the ARM code.

Reply:
I'm not sure if RISCV is always a 32-bit CPU, but on 64-bit system UNIT32_MAX+1 is just 0x100000000

Reply:
Ouch, this is indeed an error, I missed that the ARM implementation holds this value in a uint32_t and I hold it in a size_t for later comparisons on 64-bit. I've fixed this in this pull request, as this one has already merged: https://github.com/pytorch/cpuinfo/pull/199